cmd/clusterctl/cmd/version_checker.go

```diff
@@ -0,0 +1,196 @@
+// by https://github.com/cli/cli.
+func newVersionChecker(vc config.VariablesClient) *versionChecker {
+	var client *github.Client
+	token, err := vc.Get("GITHUB_TOKEN")
+	if err == nil {
+		ts := oauth2.StaticTokenSource(
+			&oauth2.Token{AccessToken: token},
+		)
+		tc := oauth2.NewClient(context.TODO(), ts)
+		client = github.NewClient(tc)
+	} else {
+		client = github.NewClient(nil)
+	}
+
+	return &versionChecker{
+		stateFilePath: filepath.Join(homedir.HomeDir(), ".cluster-api", "state.yaml"),
+		cliVersion:    version.Get,
+		githubClient:  client,
+	}
+}
+
+// ReleaseInfo stores information about the release.
+type ReleaseInfo struct {
+	Version string
+	URL     string
+}
+
+// VersionState stores the release info and the last time it was updated.
+type VersionState struct {
+	LastCheck     time.Time
+	LatestRelease ReleaseInfo
+}
+
+// Check returns a message if the current clusterctl version is less than the
+// latest available release for CAPI
+// (https://github.com/kubernetes-sigs/cluster-api). It gets the latest
+// release from github at most once during a 24 hour period and caches the
+// state by default in $HOME/.cluster-api/state.yaml. If the clusterctl
+// version is the same or greater it returns nothing.
+func (v *versionChecker) Check() (string, error) {
+	log := logf.Log
+	cliVer, err := semver.ParseTolerant(v.cliVersion().GitVersion)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to semver parse clusterctl GitVersion")
+	}
+
+	release, err := v.getLatestRelease()
+	if err != nil {
+		return "", err
+	}
+	latestVersion, err := semver.ParseTolerant(release.Version)
+	if err != nil {
+		return "", errors.Wrap(err, "unable to semver parse latest release version")
+	}
+
+	// if we are using a dirty dev build, just log it out
+	if strings.HasSuffix(cliVer.String(), "-dirty") {
+		log.V(1).Info("Using a dirty dev build of clusterctl.", "CLIVersion", cliVer.String(), "LatestGithubRelease", release.Version)
+		return "", nil
+	}
+
+	// if the cli version is a dev build off of the latest available release,
+	// the just log it out as informational.
+	if strings.HasPrefix(cliVer.String(), latestVersion.String()) && gitVersionRegEx.MatchString(cliVer.String()) {
+		log.V(1).Info("Using a dev build of latest clusterctl release.", "CLIVersion", cliVer.String(), "LatestGithubRelease", release.Version)
+		return "", nil
+	}
+
+	if latestVersion.GT(cliVer) {
+		return fmt.Sprintf(`
+New clusterctl version available: v%s -> v%s
+%s
+`, cliVer, latestVersion.String(), release.URL), nil
+	}
+	return "", nil
+}
+
+func (v *versionChecker) getLatestRelease() (*ReleaseInfo, error) {
+	vs, err := readStateFile(v.stateFilePath)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to read version state file")
+	}
+
+	// if there is no release info in the state file, pull latest release from github
+	if vs == nil {
+		release, _, err := v.githubClient.Repositories.GetLatestRelease(context.TODO(), "kubernetes-sigs", "cluster-api")
```

Comment:
When we release v0.4.0, would this return the v0.3.x or v0.4.x? We're probably going to decide if we need to support previous versions for new CLI releases, for example: will users be able to install Cluster API v0.3.x with a v0.4.x binary?

cc @fabriziopandini

Reply:
I believe this would indeed return the latest available release which would be v0.4.x. 

I guess a similar question to ask: Are we expecting users of clusterctl v0.3.x to be able to install v0.4.x CAPI? It would seem that based on the `metadata.yaml` and provider contract enforcement, we technically should be able to. So I'm assuming that users should be able to install CAPI v0.3.x with a v0.4.x clusterctl.

Reply:
@ncdc @fabriziopandini Thoughts on the scenario described above?

Reply:
I think it will make our lives easier to require that minor versions match, for the time being.

Reply:
> I think it will make our lives easier to require that minor versions match, for the time being.

I'm fine with that. I think it's probably better to start off as more restrictive rather than being open at start and then forcing restrictive behavior later on.

I'll wait for feedback from others before making any changes.

Reply:
Let's do that then!

Reply:
Coolio...so if the version of clusterctl is `v0.3.x` then this will suggest upgrades for `v0.3.y` where `y > x`. It will ignore releases with other minor versions i.e. `v0.4.*`

Reply:
donzo! see last commit.